### PR TITLE
[#1279] Implement new landing page

### PR DESCRIPTION
**Pull request by vitorbaptista**
The landing page this PR implements is http://p.smth.uk/os-mockups/. To implement it, I had to:

* Create a new route `/search`, moving the current index page to it
* Add support for the query parameters in the URL for the search page, so `/search?q=Brazil` works. This was required to make the search bar work in the new home page.
* Create a new [os-styles](https://github.com/openspending/os-styles) repository that will contain the base stylesheets and related assets that we use across OpenSpending. This avoids us having to copy/paste stuff around.
* A new API endpoint `/package/stats` via https://github.com/openspending/os-conductor/pull/22
* Lots of refactorings and moving things around

I'm getting a few issues while running `npm install`, possibly related to our Node version. I'm running 4.2.6 locally, while a few packages require >=4.3, and `os-explorer` itself requires `^4.1.0`. Do you see these issues as well? If so, we might need to upgrade the Node version as part of this.

I'd be happy to walk you through the changes if you think that'd be useful for your review, as this is a big PR.

Fixes openspending/openspending#1274

### CODE REVIEW

**Comment by akariv** [comment]
I'm trying to review it in the staging environment but the Docker image is not building - could you please take a look?

**Comment by akariv** [comment]
also it doesn't build locally - regardless of Docker:

```
Error: node_modules/os-styles/src/_variables.scss
Error: File to import not found or unreadable: osbs/variables.
       Parent style sheet: /Users/adam/code/os/openspending/repos/os-explorer/node_modules/os-styles/src/_variables.scss
        on line 1 of node_modules/os-styles/src/_variables.scss
>> @import "osbs/variables";
   ^

    at options.error (/Users/adam/code/os/openspending/repos/os-explorer/node_modules/gulp-sass/node_modules/node-sass/lib/index.js:291:26)
```

**Comment by vitorbaptista** [comment]
@akariv How did you build the image? I built simply by running `docker build .` in `os-explorer` folder and everything was fine apparently. There's a potential problem with the Dockerfile, though, in that it uses Node v7, even though we develop and test `os-explorer` with Node v4.1, so this might cause these issues.

**Comment by akariv** [comment]
The docker image had an issue - that it pulled the master branch off os-explorer and did not use the local code. So the build succeeded because it was building the old master.
I fixed that dockerfile, but the problem is still happening locally. I am using node 7 locally, so it might be related to that indeed (although the error message looks unrelated).

**Comment by vitorbaptista** [comment]
@akariv I fixed the Dockerfile. The issue was that we were missing some dependencies for `imagemin`, and were installing an incorrect version of `os-styles`. I updated to use the branch from https://github.com/openspending/os-styles/pull/1. We should then merge that PR before this, and change back to using the code from `master`.

This issue also demonstrated another problem: we aren't testing the Docker build in our CI. I added it on https://github.com/openspending/os-explorer/pull/14/commits/e9ad97e0cc37cf5c79aacc63fcc20116e6aa37da. This doubled the test time, but I think it's an acceptable drawback. WDYT?

**Comment by akariv** [comment]
Hey @vitorbaptista - I just found out that Docker cloud now has CI integrations - see here for example:
https://github.com/datahq/auth/pull/7

We could enable that for openspending so that we can see that docker still builds in PRs as well.

**Comment by vitorbaptista** [comment]
@akariv I couldn't follow the link to Docker Cloud, but I remember seeing the announcement for this functionality some time ago. It's basically a competitor to Travis, right? Do you see an advantage on using it instead of simply adding `docker build .` as an extra step in our `.travis.yml`? When I read about it I thought it was interesting to run the tests inside the actual container built, but as we're not able to do that yet, it felt overkill.

---

**TL;DR To fix the search on staging we need to add `proxy_redirect / /explorer/` to the `/explorer/` location in https://github.com/openspending/os-nginx-frontend/blob/master/server-rules.conf.**

I spent some time this afternoon debugging why this code was breaking on staging. The problem we're seeing is that when you go to https://staging.openspending.org/explorer and search in the searchbox, it redirects you to https://staging.openspending.org/explorer/search?q=foo, which is correct, but then redirects you again to https://staging.openspending.org/search/?q=foo, which doesn't exist. The culprit is the trailing slash and nginx's proxying (the added trailing slash on `/search/` gives you a hint on what's happening).

So, when a webserver sees a request like http://example.org/foo, it goes:
1. Is there a file named `foo`?
1. If not, is there a directory named `foo`?
1.1. If yes, redirect to http://example.org/foo/
1.1. If not, return a 404

`search` is a folder in `os-explorer`, so when we access `/explorer/search`, the webserver (Express in our case) notices it's actually a folder and redirects to `/explorer/search/`. The problem is that, as Express is behind a proxy, it doesn't know it's being served via a subfolder, so the requests it sees is for `/search`, which it correctly redirects to `/search/`. Notice that it's an absolute path, which is why we end up on `staging.openspending.org/search`.

We could disable this Express behaviour (see `redirect` option on http://expressjs.com/en/api.html), but it creates other issues, mainly messing up with relative paths (a relative path `assets/foo.png` on `example.org/explorer` would become `example.org/assets/foo.png`, not `example.org/explorer/assets/foo.png` that we want).

The solution is to keep the base path in the redirect, so Express redirects to `/explorer/search` instead of `/search`. This can be done with [proxy_redirect](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_redirect). This changes the redirect `Location` based on whatever rule we want. The nginx configuration would become:

```nginx
location /explorer/ {
    proxy_pass http://os-explorer:8000/;
    proxy_redirect / /explorer/;  # I'd like to remove this duplication, but I couldn't find which variable had it.
}
```

Note the difference in the `Location` header with and without `proxy_redirect`:

```shell
# Without proxy_redirect
$ curl -I http://localhost/explorer/search
HTTP/1.1 301 Moved Permanently
Server: nginx/1.10.3 (Ubuntu)
Date: Wed, 12 Jul 2017 18:39:06 GMT
Content-Type: text/html; charset=UTF-8
Content-Length: 171
Connection: keep-alive
X-Powered-By: Express
Content-Security-Policy: default-src 'self'
X-Content-Type-Options: nosniff
Location: /search/

# With proxy_redirect
$ curl -I http://localhost/explorer/search
HTTP/1.1 301 Moved Permanently
Server: nginx/1.10.3 (Ubuntu)
Date: Wed, 12 Jul 2017 18:38:49 GMT
Content-Type: text/html; charset=UTF-8
Content-Length: 171
Location: http://localhost/explorer/search/
Connection: keep-alive
X-Powered-By: Express
Content-Security-Policy: default-src 'self'
X-Content-Type-Options: nosniff
```

This change also means that any redirect to `/` would become `/explorer/`, which can have unintended side-effects (if we want to `/` if the user isn't logged in, for example). However, as the projects are developed not knowing where they live in, I don't think this will be a problem.

**Comment by vitorbaptista** [comment]
The same nginx change might be useful for other projects, specially the ones that also simply serve static files.

Back on Travis building Docker, just saw that our build was broken because Travis ran out of disk space when building our image. I re-ran the build and it ran fine. Unfortunately, Travis don't seem to define how much disk space an image has, so it'll depend on which machine our job is running at the moment. If the builds fail regularly because of this, and we can't reduce our size requirements easily, that's a good reason to build on Docker Cloud instead.

**Comment by akariv** [comment]
These are cool finds @vitorbaptista, thanks! (I didn't know most of that 😄)
Just a quick q: why can't we point to `https://staging.openspending.org/explorer/search/?q=foo` in the first place, thus avoiding all these redirects?

**Comment by vitorbaptista** [comment]
@akariv I tried that by changing the form to `<form action="search/">`, but that didn't work. It also kept the issue in that https://staging.openspending.org/explorer/search/?q=foo would work but https://staging.openspending.org/explorer/search?q=foo no, which would be surprising for the user.

We could do both: use the links with trailing slash by default, but also accept it without. However, then we have another problem: Google will consider `/search` and `/search/` as two different pages. This can only be avoided by either:
1. Using a HTTP redirect
1. Using [a rel=canonical element](https://yoast.com/rel-canonical/)

Using a `rel=canonical` element can make things more complicated, so the `nginx` fix feels like the best alternative IMHO.

**Comment by vitorbaptista** [comment]
@akariv I submitted PR https://github.com/openspending/os-nginx-frontend/pull/1 to fix this. That needs to be merged before this PR, so we can see if it fixes things on staging.

**Comment by DianaMPG** [comment]
@vitorbaptista @akariv when do you think that the new landing page can be launched?
I would prepare an annoucnement

**Comment by akariv** [comment]
@vitorbaptista seems that the changes to nginx resolved the issue (but had to test in an incognito window as apparently the redirects are cached).
